### PR TITLE
[PORT] Bot Change Detection

### PR DESCRIPTION
**Pull request by tomlm**
* Add **dialoginstance.version** property for saving version string between turns.
* Add **dialog.GetVersion()** which allows a dialog to return unique string for a configuration
* Add **DialogSet.GetVersion()** which will be different if collection of dialogs and the GetVersion() have changed.
* Add **DialogContainer.GetInternalVersion()** - which detects at the container level internal changes for children.
Add **DialogContainer.CheckForVersionChange()** - which is called on BeginDialog/ContinueDialog/ResumeDialog() to detect that a dialogContainer has changed
* Add DialogEvents.VersionChanged event which is fired when change is detected
* Implement AdaptiveDialog.GetInternalVersion() to detect triggers changes, schema changes as well as dialog changes
Unit tests, etc.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs [hunk 1]
```diff
@@ -698,6 +728,54 @@ protected async Task<RecognizerResult> OnRecognize(ActionContext actionContext,
             };
         }
 
+        protected virtual void EnsureDependenciesInstalled()
+        {
+            lock (this)
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs:733]: Is there an option other than lock(this) that could be used? Many folks consider locking like that to be problematic, as it's easy for external sources to lock and end up in a deadlock. For a good discussion of this, Jon Skeet has a lengthy discussion: 
https://jonskeet.uk/csharp/threads/lockchoice.html

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogSet.cs [hunk 2]
```diff
@@ -64,6 +66,30 @@ public IBotTelemetryClient TelemetryClient
...
+                foreach (var dialog in this._dialogs)
+                {
+                    var v = this._dialogs[dialog.Key].GetVersion();
+                    if (v != null)
+                    {
+                        sb.Append(v);
+                    }
+                }
+
+                this._version = Convert.ToBase64String(SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(sb.ToString())));
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogSet.cs:87]: I just saw this calculation over in another file. Can we consolidate to a static method that takes a stringbuilder?

Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogSet.cs:87]: there is no requirement for this to be standardized or documented.  Each class can use any technique it wants to create unique version id.

Comment (Stevenic) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogSet.cs:87]: I agree... We shouldn't add this to our API. If anything we could use an internal only class to ensure that its hidden from our API surface.

---
In reply to: [414933133](https://github.com/microsoft/botbuilder-dotnet/pull/3804#discussion_r414933133) [](ancestors = 414933133)

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogContainer.cs [hunk 3]
```diff
@@ -22,5 +25,52 @@ public virtual Dialog FindDialog(string dialogId)
         {
             return this.Dialogs.Find(dialogId);
         }
+
+        /// <summary>
+        /// GetInternalVersion - Returns internal version identifier for this container.
+        /// </summary>
+        /// <remarks>
+        /// DialogContainers detect changes of all sub-components in the container and map that to an DialogChanged event.
```
Comment (Stevenic) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogContainer.cs:33]: >DialogChanged [](start = 102, length = 13)

Should be: VersionChanged

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs [hunk 4]
```diff
@@ -5,10 +5,34 @@ namespace Microsoft.Bot.Builder.Dialogs
 {
     public class DialogEvents
     {
+        /// <summary>
+        /// Event fired when a dialog beginDialog() is called.
```
Comment (Stevenic) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs:9]: >beginDialog [](start = 38, length = 11)

Should be: BeginDialogAsync()

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs [hunk 5]
```diff
@@ -303,6 +312,28 @@ public IEnumerable<Dialog> GetDependencies()
...
+                sb.Append(JsonConvert.SerializeObject(this.Schema));
+            }
+
+            // change if triggers type/constraint change 
+            foreach (var trigger in Triggers)
+            {
+                sb.Append(trigger.GetExpression().ToString());
+            }
+
+            return Convert.ToBase64String(SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(sb.ToString())));
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs:334]: This line is in several places. Worth moving to a shared static somewhere?

**Pull request merged**